### DISCUSSION TITLE
truncated input for base64 string inputs for generation tracing

### DISCUSSION THREAD

**Discussion by JTZ18** [langfuse/langfuse#4493] [Support]
i am using `langfuse==2.53.9`
I am trying to track the generation of the llm to get accurate input and output cost token calculation for multimodal inputs like images and text. It works well when i use the `from langfuse.openai import AzureOpenAI`. But using the low level SDK results in truncation

when i use the low level sdk and use 
```python
  generation = trace.generation(
      name="answer_query",
      model=AZURE_OPENAI_DEPLOYMENT_NAME,
      model_parameters={
          "temperature": TEMPERATURE,
          "max_tokens": MAX_TOKENS
      },
      start_time=start_time,
      end_time=end_time,
      input=[{'role': "system", "content": system_prompt},
             {'role': "user", "content": [{"type": "text", "text": user_prompt}]},
             {'role': "user", "content": [
                 {"type": "image_url",
                  "image_url": {
                      "url": f"data:image/jpeg;base64,{img}"
                  }} for img in context_images
             ]}
             ],
      output=response.choices[0].message.content
  )
  generation.end()

i get the following result below in this image, where input tokens are truncated and incorrectly calculated due to truncation
```
![image](https://github.com/user-attachments/assets/8f829a8a-6641-4909-97f8-72b055001253)

when i use the `from langfuse.openai import AzureOpenAI`, the inputs are properly handled and not truncated
![image](https://github.com/user-attachments/assets/a49ec821-666e-421e-9cac-29156b78c68e)

I want to make it work in the low level SDK, any ideas how i could avoid truncation for `trace.generation`?

**Comment by jannikmaierhoefer** [comment]
Hi @JTZ18, could you try to upgrade to the latest version of our SDK (2.55.0) as we released full multi-modal support last week.

**Reply by JTZ18** [reply]
Hi @jannikmaierhoefer I've just tried `langfuse==2.55.0` now i do get a different error in the logs and now both methods dont show the 

```bash
ERROR:langfuse._task_manager.media_manager:Error processing multimodal event: Failed to execute task
WARNING:langfuse:Item exceeds size limit (size: 3789904), dropping input / output / metadata of item until it fits.
```

when using the low level sdk `trace.generation()`, i still experience the same results of incorrect token counts and truncated inputs
![image](https://github.com/user-attachments/assets/7d6fb827-e307-41f2-a97d-9a4eab980dcb)




when using the `from langfuse.openai import AzureOpenAI`, i get different results. Now the token counts are correct but the inputs are still truncated and not seen
![image](https://github.com/user-attachments/assets/f266b59f-da0e-4036-b740-dbeedc768a7c)

**Reply by jannikmaierhoefer** [reply]
Hi @JTZ18, what kind of input do you have - is it text or a media file (e.g- base64 image)?

**Reply by JTZ18** [reply]
> image)?

it's a base64 image string

**Reply by marcklingen** [reply]
Hi @JTZ18,

For multi-modal LLM calls, you need to include the token counts using the `usage` key. Tokenization does not work correctly with multi-modal traces, so we must rely on the response from OpenAI. You can find more information in the documentation here: https://langfuse.com/docs/model-usage-and-cost.

**Reply by marcklingen** [reply]
Can you share a deep link to the traces here? Would love to have a look

**Reply by JTZ18** [reply]
> Can you share a deep link to the traces here? Would love to have a look

I tried it on langfuse cloud and the multimodal base64 images are showing up perfectly.  

I apologise if i did not share earlier that im trying to get multimodal to show up at least in my input. i dont mind the images not showing up, but auto tracing the generation using `langfuse.openai` should at least not truncate my input system and text prompts.  

My self hosted version is running on docker compose langfuse v2.93.4 OSS

**Reply by hassiebp** [reply]
@JTZ18 You mentioned above that all works well with the `langfuse.openai` integration and you see failure when using the low-level SDK only? 

This sentence got me confused then, could you please summarize for which cases you see which failures?
> i dont mind the images not showing up, but auto tracing the generation using `langfuse.openai` should at least not truncate my input system and text prompts.

My understanding so far:
Langfuse Cloud - OK
Self-hosted with `langfuse.openai` integration - OK
Self-hosted with low-level SDK - FAIL

All using the Python SDK version 2.55.0

**Reply by JTZ18** [reply]
> @JTZ18 You mentioned above that all works well with the `langfuse.openai` integration and you see failure when using the low-level SDK only?
> 
> This sentence got me confused then, could you please summarize for which cases you see which failures?
> 
> > i dont mind the images not showing up, but auto tracing the generation using `langfuse.openai` should at least not truncate my input system and text prompts.
> 
> My understanding so far: Langfuse Cloud - OK Self-hosted with `langfuse.openai` integration - OK Self-hosted with low-level SDK - FAIL
> 
> All using the Python SDK version 2.55.0

Apologies, let me summarise what works and what doesnt. 
Langfuse Cloud - OK (SDK 2.55.0)
Self-hosted with langfuse.openai integration - FAIL (SDK 2.55.0)
Self-hosted with langfuse.openai integration - OK (SDK 2.53.9)

Self-hosted with low-level SDK - FAIL (SDK 2.53.9)
Self-hosted with low-level SDK - FAIL (SDK 2.55.0)

For the fail cases, when theres base64 images in my messages, my input will be truncated

**Reply by hassiebp** [reply]
If all works well with the 2.55.0 SDK on Langfuse Cloud I'm suprised about the truncation messages your are seeing with the same SDK on the self-hosted instances. All truncation / media management happens SDK side. Could you please verify you are indeed running 2.55.0 against your self hosted instances, and could you please share the exact truncation messages you are seeing in your self-hosted instances? (Assuming you have also verified that all works correct on Langfuse Cloud for 2.55.0)
